# Note

### Key difference between Z-score method and Windorization

Both Winsorization and the Z-score method deal with outliers and are often used during data preprocessing to reduce their impact.

**Z-score method**

* Detects outliers

* Based on how many standard deviations a value is from the mean

* Formula:

            $$z = \frac{x - \mu}{\sigma}$$

* Typical rule: values with ∣z∣>3 are considered outliers

* What happens to outliers?

    * Usually removed or flagged

    * Sometimes capped, but that’s not the core idea


**Winsorization**

* Modifies outliers instead of removing them

* Extreme values are replaced with boundary values (percentiles)

* Example:

    * Bottom 1% → replaced with 1st percentile value

    * Top 1% → replaced with 99th percentile value

* Keeps dataset size the same

* Reduces the influence of extreme values on mean & variance

### Handling Duplicate Data

Removing duplicate data from a dataset is essential because duplicate entries can negatively impact the performance of machine learning algorithms. If not addressed, duplicates may introduce bias, distort model training, and lead to inaccurate predictions. Additionally, redundant data increases the overall dataset size unnecessarily, which can slow down processing time and reduce computational efficiency.



In [1]:
import pandas as pd

In [2]:
data = {"name": ["a","b","c","d","a","c"], "eng":[8, 7, 5, 8, 8, 4], "maths": [2, 3, 4, 5, 2, 4]}    # Here, a contains duplicate data but c doesn't

In [3]:
# Let's create a dataset
df = pd.DataFrame(data)

In [4]:
df

,name,eng,maths
0,a,8,2
1,b,7,3
2,c,5,4
3,d,8,5
4,a,8,2
5,c,4,4


In [5]:
df.duplicated()

0    False
1    False
2    False
3    False
4     True
5    False
dtype: bool

Here, we can clearly observe that the entries at index 0 and 4 contain identical values, whereas the entries at index 2 and 5 differ from each other. Therefore, duplicate data is present only at indices 0 and 4.

In [6]:
# Let's drop the duplicate columns
df.drop_duplicates(keep='first', inplace=True)

The argument `keep='first'` lets you keep the first duplicate value and removes the rest

In [7]:
df

,name,eng,maths
0,a,8,2
1,b,7,3
2,c,5,4
3,d,8,5
5,c,4,4


Now let's apply the method on a real dataset

In [8]:
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import kagglehub
from kagglehub import KaggleDatasetAdapter
import urllib.request
%matplotlib inline

warnings.filterwarnings('ignore')    # This can be used for removing any future warnings

In [9]:
# pip install --upgrade kagglehub

In [10]:
# pip install --upgrade pip

In [11]:
# Set the path to the file you'd like to load
file_path = "loan.csv"  # file inside tanishaj225/loancsv dataset

# Load dataset into pandas DataFrame
loan_df = kagglehub.load_dataset(
    KaggleDatasetAdapter.PANDAS,
    "tanishaj225/loancsv",   # dataset slug
    path=file_path           # specify file to load
)


In [12]:
loan_df.head()

,Loan_ID,Gender,Married,Dependents,Education,Self_Employed,ApplicantIncome,CoapplicantIncome,LoanAmount,Loan_Amount_Term,Credit_History,Property_Area,Loan_Status
0,LP001002,Male,No,0,Graduate,No,5849,0.0,NaN,360.0,1.0,Urban,Y
1,LP001003,Male,Yes,1,Graduate,No,4583,1508.0,128.0,360.0,1.0,Rural,N
2,LP001005,Male,Yes,0,Graduate,Yes,3000,0.0,66.0,360.0,1.0,Urban,Y
3,LP001006,Male,Yes,0,Not Graduate,No,2583,2358.0,120.0,360.0,1.0,Urban,Y
4,LP001008,Male,No,0,Graduate,No,6000,0.0,141.0,360.0,1.0,Urban,Y


In [13]:
loan_df.describe()

,ApplicantIncome,CoapplicantIncome,LoanAmount,Loan_Amount_Term,Credit_History
count,614.000000,614.000000,592.000000,600.00000,564.000000
mean,5403.459283,1621.245798,146.412162,342.00000,0.842199
std,6109.041673,2926.248369,85.587325,65.12041,0.364878
min,150.000000,0.000000,9.000000,12.00000,0.000000
25%,2877.500000,0.000000,100.000000,360.00000,1.000000
50%,3812.500000,1188.500000,128.000000,360.00000,1.000000
75%,5795.000000,2297.250000,168.000000,360.00000,1.000000
max,81000.000000,41667.000000,700.000000,480.00000,1.000000


In [14]:
loan_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            614 non-null    object 
 1   Gender             601 non-null    object 
 2   Married            611 non-null    object 
 3   Dependents         599 non-null    object 
 4   Education          614 non-null    object 
 5   Self_Employed      582 non-null    object 
 6   ApplicantIncome    614 non-null    int64  
 7   CoapplicantIncome  614 non-null    float64
 8   LoanAmount         592 non-null    float64
 9   Loan_Amount_Term   600 non-null    float64
 10  Credit_History     564 non-null    float64
 11  Property_Area      614 non-null    object 
 12  Loan_Status        614 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 62.5+ KB


In [15]:
loan_df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
609    False
610    False
611    False
612    False
613    False
Length: 614, dtype: bool

In [16]:
loan_df.duplicated().unique()

array([False])

To check for duplicate values, I used `.duplicated().unique()` to identify the distinct boolean results returned by the `.duplicated()` method. Since this method produces a boolean output (`True` for duplicate rows and `False` otherwise), we expect only two possible values. In this case, `.unique()` did not return any `True` values, which indicates that no duplicate records are present in the dataset.

Another way to verify duplicates is by using `loan_df.drop_duplicates(keep='first', inplace=True)` and then checking whether the dataset’s shape changes. If the number of rows decreases, it confirms that duplicate entries were present and have been removed. If the shape remains the same, it indicates that no duplicates existed.

In [17]:
# Or we can also use .sum instead of .unique
loan_df.duplicated().sum()

np.int64(0)

### Checking vs Removing Duplicates

It is better to **check for duplicate values** first instead of removing them immediately. This helps us understand whether duplicates actually exist before modifying the dataset.

Using:

***loan_df.duplicated().unique()***


is a safe way to check if duplicates are present. Since `.duplicated()` returns boolean values (`True` for duplicates and `False` for unique rows), if `True` does not appear, it means there are no duplicate rows.

However, a more practical way to check duplicates is:

***loan_df.duplicated().sum()***


This directly tells us the number of duplicate rows:

* If the result is `0`, there are no duplicates.

* If the result is greater than `0`, duplicates are present.

After confirming duplicates exist, we can remove them using:

***loan_df.drop_duplicates(keep='first', inplace=True)***


Checking first is considered a good data practice because it avoids modifying the dataset blindly and helps maintain transparency in the data cleaning process.

---

One last thing before closing this note.
Do you remember the heavy right skewed Co-Applicant Income let me show you the description of it

|        |ApplicantIncome |CoapplicantIncome | LoanAmount | Loan_Amount_Term |Credit_History |
|--------|----------------|------------------|------------|------------------|---------------|
| count  | 614.000000     | 614.000000       | 592.000000 | 600.00000        | 564.000000    |
| mean   | 5403.459283    | 1621.245798      | 146.412162 | 342.00000        | 0.842199      |
| std    | 6109.041673    | 2926.248369      | 85.587325  | 65.12041         | 0.364878      |
| min    | 150.000000     | 0.000000         | 9.000000   | 12.00000         | 0.000000      |
| 25%    | 2877.500000    | 0.000000         | 100.000000 | 360.00000        | 1.000000      |
| 50%    | 3812.500000    | 1188.500000      | 128.000000 | 360.00000        | 1.000000      |
| 75%    | 5795.000000    | 2297.250000      | 168.000000 | 360.00000        | 1.000000      |
| max    | 81000.000000   | 41667.000000     | 700.000000 | 480.00000        | 1.000000      |



## CoapplicantIncome – Outliers, Z-score & Winsorization Summary

#### Distribution Diagnosis

From the statistics:

* Mean ≈ 1621

* Std ≈ 2926 (very high)

* 25% = 0

* Median = 1188

* 75% = 2297

* Max = 41667

Interpretation:

* Many values are 0 (no co-applicant)

* A few very large incomes stretch the mean

* Distribution is highly right-skewed

* This is not normally distributed

* This is structural skew, not noise.

#### Why Z-Score Removed 67%

Z-score assumes:

##### *Data ∼ Normal Distribution*

But your data is:

* Zero-inflated

* Right-skewed

* Heavy-tailed

* So when applying:

  **∣𝑧∣>3**

It incorrectly flags many normal values as outliers.

Problem is **misuse of Z-score**, not the data itself.

#### Winsorization Concern Clarified

Winsorization:

* Caps only extreme percentiles

* Does NOT replace all large values with one single number

* Only affects the chosen tail %

Example:

* Winsorize at 95%

* Only top 5% capped

* 95% unchanged

So 67% will NOT collapse into the same value unless you cap 67%.

#### When Winsorization Works Well

* Rare but extreme anomalies
* Symmetric distribution with few tail errors
* Linear models sensitive to extreme values

***Winsorization = robustness tool***

#### When Winsorization Is NOT Ideal

* Highly skewed data
* Zero-inflated distributions
* Financial features with natural long tails

It compresses real signal instead of fixing noise.

### Better Approach for CoapplicantIncome

* Log Transform (Best choice)

        np.log1p(CoapplicantIncome)

Why?

* Compresses large values

* Preserves ranking

* Keeps variance structure

* No artificial duplicates

#### Or leave as-it-is for tree models

Models like:

* Random Forest

* XGBoost

* LightGBM

are naturally robust to skewness.

#### Key Takeaways

* High variance ≠ bad feature

* Z-score fails on skewed data

* Winsorization is for extreme anomalies, not skew correction

* Log transform > Winsorization for skewed financial features

## Final Understanding



Winsorization is useful for extreme anomalies, but not appropriate for naturally skewed financial features like CoapplicantIncome.

---

## TL;DR

***CoapplicantIncome – Key Insight***

* Distribution is zero-inflated + highly right-skewed (not normal).

* Z-score failed because it assumes normal distribution → falsely flagged 67% as outliers.

* Winsorization only caps extreme tails — it does not collapse 67% unless capped that aggressively.

* Problem isn’t outliers — it’s natural skew.

* Best fix: **Log transform** (`log1p`), or leave as-it-is for tree models.

* Winsorization = robustness tool, not skewness correction tool.